File: extensions/vscode/src/auth/AuthService.ts
import type { AiorgClient } from '../api/AiorgClient';

const BEARER_KEY = 'aiorg.bearer';

export interface AuthState {
  signedIn: boolean;
}

interface SecretStorageLike {
  get(key: string): Promise<string | undefined>;
  store(key: string, value: string): Promise<void>;
  delete(key: string): Promise<void>;
}

export interface AuthServiceDeps {
  secrets: SecretStorageLike;
  client: AiorgClient;
  openBrowser: (url: string) => Promise<void>;
  serverUrl: string;
}

export class AuthService {
  private pendingState: string | null = null;
  private listeners = new Set<(state: AuthState) => void>();

  constructor(private readonly deps: AuthServiceDeps) {}

  onAuthChanged(listener: (state: AuthState) => void): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  async getToken(): Promise<string | null> {
    return (await this.deps.secrets.get(BEARER_KEY)) ?? null;
  }

  async isSignedIn(): Promise<boolean> {
    return (await this.getToken()) !== null;
  }

  async startSignIn(): Promise<void> {
    const state = this.randomState();
    this.pendingState = state;

    const url = `${this.deps.serverUrl}/auth/slack?return=ide&state=${encodeURIComponent(state)}`;
    await this.deps.openBrowser(url);
  }

  async completeSignIn(payload: { token: string; state: string }): Promise<void> {
    if (this.pendingState === null || payload.state !== this.pendingState) {
      throw new Error('sign-in state mismatch');
    }

    this.pendingState = null;

    const result = await this.deps.client.post<{ bearer: string }>(
      '/api/ide/auth/exchange',
      payload,
      { authenticated: false },
    );

    await this.deps.secrets.store(BEARER_KEY, result.bearer);
    this.fire({ signedIn: true });
  }

  async signOut(): Promise<void> {
    try {
      await this.deps.client.post('/api/ide/auth/revoke', {});
    } catch {
      // Server-side revoke is best-effort; the client-side clear is what matters.
    }

    await this.deps.secrets.delete(BEARER_KEY);
    this.fire({ signedIn: false });
  }

  async handleUnauthorized(): Promise<void> {
    await this.deps.secrets.delete(BEARER_KEY);
    this.fire({ signedIn: false });
  }

  pendingStateForTest(): string | null {
    return this.pendingState;
  }

  private randomState(): string {
    const bytes = new Uint8Array(32);
    crypto.getRandomValues(bytes);
    return Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
  }

  private fire(state: AuthState): void {
    for (const listener of this.listeners) {
      try {
        listener(state);
      } catch {
        // listeners shouldn't crash other listeners
      }
    }
  }
}

File: extensions/jetbrains/src/main/kotlin/app/tokenslayer/auth/AuthService.kt
// AuthService.kt
package app.tokenslayer.auth

import app.tokenslayer.api.TokenSlayerClient
import com.google.gson.JsonObject
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.security.SecureRandom

class AuthService(
    private val secrets: SecretStore,
    private val client: TokenSlayerClient,
    private val openBrowser: (String) -> Unit,
    private val serverUrl: () -> String,
    private val loopback: LoopbackServer = JdkLoopbackServer(),
) {
    private var pendingState: String? = null
    private val listeners = mutableSetOf<(Boolean) -> Unit>()
    private val rng = SecureRandom()

    fun onAuthChanged(listener: (Boolean) -> Unit): () -> Unit {
        listeners.add(listener)
        return { listeners.remove(listener) }
    }

    fun token(): String? = secrets.get()
    fun isSignedIn(): Boolean = secrets.get() != null

    fun startSignIn() {
        val state = randomHex(32)
        pendingState = state
        val port = loopback.start { token, callbackState ->
            completeSignIn(token, callbackState)
        }
        val redirect = URLEncoder.encode("http://127.0.0.1:$port/callback", StandardCharsets.UTF_8)
        openBrowser("${serverUrl()}/auth/slack?return=ide&client=jetbrains&state=$state&redirect=$redirect")
    }

    fun completeSignIn(token: String, state: String) {
        val pending = pendingState
        if (pending == null || state != pending) error("sign-in state mismatch")
        pendingState = null
        val body = JsonObject().apply {
            addProperty("token", token)
            addProperty("state", state)
        }
        val result = client.postJson("/api/ide/auth/exchange", body, authenticated = false)
        secrets.set(result.get("bearer").asString)
        fire(true)
    }

    fun signOut() {
        try {
            client.postJson("/api/ide/auth/revoke", JsonObject())
        } catch (_: Exception) {
            // best-effort; client-side clear is what matters
        }
        secrets.clear()
        fire(false)
    }

    fun handleUnauthorized() {
        secrets.clear()
        fire(false)
    }

    fun pendingState(): String? = pendingState

    private fun fire(signedIn: Boolean) {
        for (l in listeners.toList()) {
            try { l(signedIn) } catch (_: Exception) {}
        }
    }

    private fun randomHex(bytes: Int): String {
        val b = ByteArray(bytes); rng.nextBytes(b)
        return b.joinToString("") { "%02x".format(it) }
    }
}
